add post about using erlang releases without epmd (#1)
erlware/blog.erlware.org: content/epmdlessless.md

+++
author = "Tristan Sloughter"
type="post"
categories = ["Erlang"]
date = 2020-11-30T10:41:00Z
description = ""
draft = false
slug = "epmdlessless"
tags = ["Erlang"]
title = "Running Erlang Releases without EPMD on OTP 23.1+"
+++

> With Erlang/OTP 23.1 and relx 4.2 (included in rebar3 3.14.3) it is now possible
> to conveniently run Erlang releases without relying on EPMD or any third party
> dependency, like [epmdless](https://github.com/tsloughter/epmdless), for opening a
> remote shell or clustering nodes.

Reviewer: The sentence fragment with epmdless is a bit odd there. The explanation about EPMD isn't given until later in the text as well. I'm thinking it might make sense to turn this into an intro.

```suggestion
Erlang/OTP deployments that want to provide shell access relied on something
called the Erlang Port Mapper Daemon (EPMD), a separate process that handled
all distribution mechanisms. This often caused issues where local shell access to
a running production system forced the use of peer-to-peer mechanisms that didn't
always work well in all environments. Previous attempts at working around this
requirement used third party dependencies such as
[epmdless](https://github.com/tsloughter/epmdless), but still weren't ideal.

With Erlang/OTP 23.1 and relx 4.2 (included in Rebar3 3.14.3) it is now possible
to conveniently run Erlang releases without relying on EPMD for opening a
remote shell or clustering nodes. This is particularly going to make Erlang releases
simpler to configure in locked down environments, and easier to use with containers.
```

Author: Merging this but then changing it a bit.

I'm not sure what you mean by peer-to-peer mechanisms? Is that referring to the remote shell also being a full node listening on a distribution port or to something around how ports were shared between physical nodes in EPMD?

Reviewer: yeah. It's not obvious to people sometimes why just getting a REPL involves the same machinery as setting up a mesh cluster.

> The Erlang Port Mapper Daemon (EPMD) is a separate process that is started, if

Reviewer: ```suggestion
EPMD is a separate process that is started, if
```
moved the acronym definition in my previous suggestion; if accepted, it has become redundant here.

not already running, when an Erlang node boots with distribution enabled --
meaning `-sname` or `-name` is passed to the boot command. The new node [talks to
the EPMD process](http://erlang.org/doc/apps/erts/erl_dist_protocol.html#epmd-protocol) to register its name with a port that the node will also bind

> to. When a node wants to connect to another it must lookup the port the other
> node is listening on and does so through EPMD. This is required both for

Reviewer: ```suggestion
to. When a node wants to connect to another one it must look up the port the other
node is listening on through EPMD. This is required both for
```

clustering nodes together and connecting a remote shell on the command line --
which is essentially the same as clustering because a new node is started before
attaching to the remote node to bring up a shell on that node.

> This has been made possible through a couple new options added to `erl`:

Reviewer: ```suggestion
Bypassing EPMD has been made possible through a couple new options added to `erl`:
```

* `-erl_epmd_port P`: Configures the port this node will listen on and use when
  connecting to remote nodes.

> * `-dist_listen false`: Setting this keeps the node from binding to the port,
>   instead it is only used for connecting to remote nodes. The use case for this

Reviewer: ```suggestion
* `-dist_listen false`: Setting this keeps the node from binding to the port, prevents it from acting as a server,
 and instead uses the port only for connecting to remote nodes. The use case for this
```
trying to be explicit in explaining the end goal of not binding to the port. Technically I guess we bind on it when connecting, but we don't listen/accept on it?

Author: Right, I only ever use "bind" to refer to a server listening. If its called "binding" when a client opens a connection it still isn't on that server port but some other ephemeral port, right?

Reviewer: right. So you still have to bind on some socket, but it's not the same exact mechanism as listening/accepting protocol-wise.

  is when opening a remote shell `-remsh <node>` (`remote_console` in the
  release script) to a local node.

> And one for [`erl_call`](http://erlang.org/doc/man/erl_call.html) :

Reviewer: ```suggestion
An extra option has also been added to [`erl_call`](http://erlang.org/doc/man/erl_call.html) :
```
People will have lost context after reading the two options, so "one" -> "option" being explicit may be helpful

* `-address P`: Sets the address and/or port to connect to the running Erlang

>   node on.

Reviewer: ```suggestion
  node.
```

  
As of OTP 23.1+ the `relx` generated release script uses `erl_call` instead of
the custom `escript` `nodetool` when the commands `rpc` and `eval` are used to
run code on a running node.

These options are used automatically by the release shell script when starting a
node, opening a remote shell or interacting with the running node through `rpc`
and `eval`. The only change you as a user must make to run your release without
`epmd` is to set the OS environment variable `ERL_DIST_PORT`.

When `ERL_DIST_PORT` is set in the environment the option to not start EPMD
(`-start_epmd false`) is added to the `erl` arguments automatically, along with
`-erl_epmd_port ${ERL_DIST_PORT}`. The `ERL_DIST_PORT` is also used by
`erl_call` with `-address ${ERL_DIST_PORT}`.

[epmdlessless](https://github.com/tsloughter/epmdlessless) is an example project

> that comes with a [docker-compose.yml]() setup that makes it simple to play

Reviewer: link missing

```
$ docker exec -ti node_c bin/epmdlessless remote_console
Erlang/OTP 23 [erts-11.1.3] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [hipe]

Eshell V11.1.3  (abort with ^G)
(epmdlessless@node_c)1> net_kernel:connect_node(epmdlessless@node_b).
true
(epmdlessless@node_c)2> erlang:nodes().
[epmdlessless@node_b,epmdlessless@node_a]
```

Test the `rpc` command by running `erlang:nodes()` on `node_b`:

```
$ docker exec -ti node_b bin/epmdlessless rpc erlang nodes
[epmdlessless@node_a, epmdlessless@node_c]
```

Lastly, you can verify that none of the commands caused EPMD to be run:

```
$ docker exec -ti node_b erts-11.1.3/bin/epmd -names
epmd: Cannot connect to local epmd
```

## Separate Ports: Where epmdless is Still Needed

It is important to note that this requires every node to listen on the same
port, which requires every node to have its own IP. If this isn't an option, or
for whatever reason you need to use separate ports for nodes, then [epmdless](https://github.com/tsloughter/epmdless)
is an option to replace EPMD and track the node to port mapping. But if you are
running with containers in an environment like Kubernetes you can now keep EPMD

> from running in each container by simply upgrading Erlang and rebar3, and

Reviewer: ```suggestion
from running in each container by simply upgrading Erlang and Rebar3, and
```
